### ISSUE TITLE
Object files in the core/src directory interfere with CMake builds

### ISSUE THREAD

**Issue by DavidPoliakoff** [kokkos/kokkos#2298]
Hey folks, 

I'm trying to build `develop` or `master` on my Mac with a Clang 9.0.0 install without OpenMP enabled. I'm using CMake, and look at the cache variables, and see 

```
 Kokkos_ENABLE_PTHREAD            OFF
 Kokkos_ENABLE_QTHREAD            OFF
```

That said, when I try to link an application against it, I get a bunch of linker errors,

```
clang++ -o madebad madebad.cpp -I /Library/Developer/CommandLineTools/usr/include/c++/v1/ -isystem /Library/Developer/CommandLineTools/usr/include/c++/v1/  -isystem /usr/include -fsanitize=address -fsanitize-recover=address -std=c++11 -I/Users/.../install/kokkos/include -L/Users/.../install/kokkos/lib -lkokkos -ldl -fPIC --gcc-toolchain=/usr/bin

Undefined symbols for architecture x86_64:
  "Kokkos::Impl::ThreadsExec::initialize(unsigned int, unsigned int, unsigned int, bool)", referenced from:
      Kokkos::Threads::initialize(unsigned int, unsigned int, unsigned int, bool) in libkokkos.a(Kokkos_Core.cpp.o)
  "Kokkos::Impl::ThreadsExec::is_initialized()", referenced from:
      Kokkos::Threads::is_initialized() in libkokkos.a(Kokkos_Core.cpp.o)
  "Kokkos::Impl::ThreadsExec::print_configuration(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, bool)", referenced from:
      Kokkos::Threads::print_configuration(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, bool) in libkokkos.a(Kokkos_Core.cpp.o)
  "Kokkos::Impl::ThreadsExec::fence()", referenced from:
      Kokkos::Threads::impl_static_fence() in libkokkos.a(Kokkos_Core.cpp.o)
      Kokkos::Impl::ParallelFor<Kokkos::Impl::hostspace_parallel_deepcopy(void*, void const*, long)::$_0, Kokkos::RangePolicy<Kokkos::Threads>, Kokkos::Threads>::execute() const in libkokkos.a(Kokkos_HostSpace_deepcopy.cpp.o)
      Kokkos::Impl::ParallelFor<Kokkos::Impl::hostspace_parallel_deepcopy(void*, void const*, long)::$_1, Kokkos::RangePolicy<Kokkos::Threads>, Kokkos::Threads>::execute() const in libkokkos.a(Kokkos_HostSpace_deepcopy.cpp.o)
      Kokkos::Impl::ParallelFor<Kokkos::Impl::hostspace_parallel_deepcopy(void*, void const*, long)::$_2, Kokkos::RangePolicy<Kokkos::Threads>, Kokkos::Threads>::execute() const in libkokkos.a(Kokkos_HostSpace_deepcopy.cpp.o)
  "Kokkos::Impl::ThreadsExec::start(void (*)(Kokkos::Impl::ThreadsExec&, void const*), void const*)", referenced from:
      Kokkos::Impl::ParallelFor<Kokkos::Impl::hostspace_parallel_deepcopy(void*, void const*, long)::$_0, Kokkos::RangePolicy<Kokkos::Threads>, Kokkos::Threads>::execute() const in libkokkos.a(Kokkos_HostSpace_deepcopy.cpp.o)
      Kokkos::Impl::ParallelFor<Kokkos::Impl::hostspace_parallel_deepcopy(void*, void const*, long)::$_1, Kokkos::RangePolicy<Kokkos::Threads>, Kokkos::Threads>::execute() const in libkokkos.a(Kokkos_HostSpace_deepcopy.cpp.o)
      Kokkos::Impl::ParallelFor<Kokkos::Impl::hostspace_parallel_deepcopy(void*, void const*, long)::$_2, Kokkos::RangePolicy<Kokkos::Threads>, Kokkos::Threads>::execute() const in libkokkos.a(Kokkos_HostSpace_deepcopy.cpp.o)
  "Kokkos::Impl::ThreadsExec::finalize()", referenced from:
      Kokkos::Threads::finalize() in libkokkos.a(Kokkos_Core.cpp.o)
  "Kokkos::Serial::is_initialized()", referenced from:
      Kokkos::View<double*, Kokkos::HostSpace>::View<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(Kokkos::Impl::ViewCtorProp<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > const&, std::__1::enable_if<!(Impl::ViewCtorProp<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::has_pointer), Kokkos::LayoutRight>::type const&) in madebad-0c3262.o
  "Kokkos::Threads::concurrency()", referenced from:
      Kokkos::Impl::hostspace_parallel_deepcopy(void*, void const*, long) in libkokkos.a(Kokkos_HostSpace_deepcopy.cpp.o)
      Kokkos::RangePolicy<Kokkos::Threads>::set_auto_chunk_size() in libkokkos.a(Kokkos_HostSpace_deepcopy.cpp.o)
```

Any thoughts?

**Comment by dalg24** [comment]
`master` and `develop` are quite different at that time since we merged #2104 

Please use `develop` and send details about how you configured with CMake.

**Comment by DavidPoliakoff** [comment]
With `develop` I configured

`cmake -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_INSTALL_PREFIX=[install_prefix]`

**Comment by crtrott** [comment]
Something is very wrong with your link. It shouldn't really be possible to build without Serial::is_initialized ..

**Comment by DavidPoliakoff** [comment]
@crtrott, yeah, it looks like the files are getting built weirdly, I look at the object files

```
nm ./core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Serial.cpp.o
0000000000000000 T __Z46KOKKOS_CORE_SRC_IMPL_SERIAL_PREVENT_LINK_ERRORv
```

And that's it. I don't know why, though, the build looks normal when I run `make`

```
.../clang++   -I/.../src/kokkos-clean/build/core/src -I/.../src/kokkos-clean/core/src -I/.../src/kokkos-clean/build  -std=gnu++11 -o CMakeFiles/kokkoscore.dir/impl/Kokkos_Serial.cpp.o -c /.../src/kokkos-clean/core/src/impl/Kokkos_Serial.cpp
```

**Comment by crtrott** [comment]
This means you didn't have any backends enabled maybe?

**Comment by DavidPoliakoff** [comment]
```
-- Setting policy CMP0074 to use <Package>_ROOT variables
-- The project name is: Kokkos
-- Setting Kokkos_ENABLE_SERIAL=ON from Kokkos_DEVICES
-- Setting default Kokkos CXX standard to 11
-- Using -std=gnu++11 for C++11 extensions as feature
-- Execution Spaces:
--   Device Parallel: None
--     Host Parallel: NONE
--       Host Serial: SERIAL
--
-- Architectures:
-- Configuring done
-- Generating done
-- Build files have been written to: /.../src/kokkos-clean/build
```

?

**Comment by dalg24** [comment]
@DavidPoliakoff Sorry but I am not able to reproduce.  Are you on Slack?

**Comment by DavidPoliakoff** [comment]
Okay, this one turned out to be weird. As it turns out, I had typed `make` in the `core/src` directory. Apparently, the existence of `.o` files in that directory changes how your CMake builds work. This raises a couple of possibilities

1. if we can make that behavior not happen, I'd be happy
2. if the behavior exists for legacy reasons, I think a warning in big angry letters should be emitted

If you're stumbling onto this thread via a search, go into `core/src` and type `make clean`, then rebuild in your cmake directory

**Comment by jjwilke** [comment]
I am not able to reproduce this. What CMake version?

**Comment by DavidPoliakoff** [comment]
@jjwilke : back at the computer

On my Mac

`cmake version 3.15.20190730-g0e2e3e0`

On Blake

`cmake version 3.10.2`

The critical thing is that the in-source build uses the Makefile, not the Cmake system

**Comment by DavidPoliakoff** [comment]
Sorry for the spam, I just investigated this more. An in-source _CMake_ build has different effects. It doesn't prevent an out-of-source CMake build from functioning, but it does override the Makefile we ship with Kokkos. I'd recommend merging my PR (or the custom_target variant I discuss with @dalg24 ) and either or both of

1. Removing the built-in Makefile
2. Merging #2303 

Another path would be to look at why object files in the source tree interfere with out-of-source builds, but I don't have the cycles to chase that down

**Comment by Rombur** [comment]
Why do we still have our own `Makefile`? If I understand the problem correctly,  the problem happens because we have two build systems (`Makefile` and `CMake`) and using `Makefile` will trip `CMake`. Removing the `Makefile` will fix your problem, correct?

**Comment by DavidPoliakoff** [comment]
Absolutely, removing the Makefile would fix our problem. I'm worried the Makefile exists for Legacy Reasons, and is not removable. If removing the Makefile is an option I'm all for it.

**Comment by ibaned** [comment]
Closing in favor of #2310

**Comment by DavidPoliakoff** [comment]
Fixed via #2339
